ledger/error/generator/lib/src/main/scala/com/daml/error/generator/ErrorCodeDocumentationGenerator.scala

```diff
@@ -8,239 +8,229 @@ import java.lang.reflect.Modifier
     errorCodes.view.map(_.id).groupBy(identity).collect {
       case (code, occurrences) if occurrences.size > 1 =>
         sys.error(
-          s"Error code $code is used ${occurrences.size} times but we require each error code to be unique! " +
-            s"Make these error code unique to make this assertion run through"
+          s"Error code $code is used ${occurrences.size} times but we require each error code to be unique!"
         )
     }
+    errorCodes
+      .map { errorCode =>
+        val annotations = parseErrorCodeAnnotations(errorCode)
+        ErrorCodeDocItem(
+          className = errorCode.getClass.getName,
+          category = simpleClassName(errorCode.category),
+          hierarchicalGrouping = errorCode.parent,
+          conveyance = errorCode.errorConveyanceDocString,
+          code = errorCode.id,
+          deprecation = annotations.deprecation,
+          explanation = annotations.explanation,
+          resolution = annotations.resolution,
+        )
+      }
+      .sortBy(_.code)
+  }
 
-    val groups = getInstances[ErrorGroup]
-    groups.view.map(_.errorClass).groupBy(identity).collect {
+  def getErrorGroupItems(
+      searchPackagePrefixes: Array[String] = DefaultPackagePrefixes
+  ): Seq[ErrorGroupDocItem] = {
+    val errorGroups = findInstancesOf[ErrorGroup](searchPackagePrefixes)
+    errorGroups.view.map(_.errorClass).groupBy(identity).collect {
       case (group, occurrences) if occurrences.size > 1 =>
         sys.error(
-          s"There are ${occurrences.size} groups named $group but we require each group class name to be unique! " +
-            s"Make these group class namers unique to make this assertion run through"
+          s"There are ${occurrences.size} groups named $group but we require each group class name to be unique! "
         )
     }
-
-    (errorCodes.map(convertToDocItem).sortBy(_.code), groups.map(convertToDocItem))
+    errorGroups.map { errorGroup =>
+      ErrorGroupDocItem(
+        errorClass = errorGroup.errorClass,
+        className = errorGroup.fullClassName,
+        explanation = parseErrorGroupAnnotations(errorGroup).explanation,
+      )
+    }
   }
 
-  private def getInstances[T: ru.TypeTag]: Seq[T] =
-    new Reflections(prefixes)
-      .getSubTypesOf(runtimeMirror.runtimeClass(typeOf[T]))
-      .asScala
-      .view
-      .collect {
-        case clazz if !Modifier.isAbstract(clazz.getModifiers) =>
-          clazz.getDeclaredField("MODULE$").get(clazz).asInstanceOf[T]
+  def getErrorCategoryItem(errorCategory: ErrorCategory): ErrorCategoryDocItem = {
+    val mirroredType = runtimeMirror.reflect(errorCategory)
+    val annotations: Seq[ru.Annotation] = mirroredType.symbol.annotations
+    val description = new SettableOnce[String]
+    val resolution = new SettableOnce[String]
+    val retryStrategy = new SettableOnce[String]
+    annotations.foreach { annotation =>
+      lazy val parsedValue = parseAnnotationValue(annotation.tree)
```

Comment:
You can achieve the same without implicit locks by inlining the call where it's used. It's a few characters, I wouldn't mind it.

Reply:
Ok